include/audio/dmic.h

```diff
@@ -0,0 +1,301 @@
+	/* minimum duty cycle in % supported by the mic */
+	u8_t	min_pdm_clk_dc;
+	/* maximum duty cycle in % supported by the mic */
+	u8_t	max_pdm_clk_dc;
+
+	/* parameters unique to each PDM controller */
+
+	/* Bit mask to optionally invert PDM clock */
+	u8_t	pdm_clk_pol;
+	/* Bit mask to optionally invert mic data */
+	u8_t	pdm_data_pol;
+	/* Collection of clock skew values for each PDM port */
+	u32_t	pdm_clk_skew;
+};
+
+/**
+ * Configuration of the PCM streams to be output by the PDM hardware
+ */
+struct pcm_stream_cfg {
+	/*
+	 * if either rate or width is set to 0 for a stream,
+	 * the stream would be disabled
+	 */
+
+	/* PCM sample rate of stream */
+	u32_t			pcm_rate;
+	/* PCM sample width of stream */
+	u8_t			pcm_width;
+	u8_t			reserved;
+	/* PCM sample block size per transfer */
+	u16_t			block_size;
+	/* SLAB for DMIC driver to allocate buffers for stream */
+	struct k_mem_slab	*mem_slab;
+};
+
+/**
+ * Mapping/ordering of the PDM channels to logical PCM output channel
+ */
+struct pdm_chan_cfg {
+	/*
+	 * mapping of PDM controller and mic channel to logical channel
+	 * since each controller can have 2 audio channels (stereo),
+	 * there can be total of 8x2=16 channels.
+	 * The actual number of channels shall be described in
+	 * pcm_stream_cfg.num_chan.
+	 * if 2 streams are enabled, the channel order will be the same for
+	 * both streams
+	 * Each channel is described as a 4 bit number, the least significant
+	 * bit indicates LEFT/RIGHT selection of the PDM controller.
+	 * The most significant 3 bits indicate the PDM controller number.
+	 *	bits 0-3 are for channel 0, bit 0 indicates LEFT or RIGHT
+	 *	bits 4-7 are for channel 1, bit 4 indicates LEFT or RIGHT
+	 *	and so on.
+	 * CONSTRAINT: The LEFT and RIGHT channels of EACH PDM controller needs
+	 * to be adjacent to each other.
+	 */
+	/* Requested channel map */
+	u32_t	req_chan_map_lo;	/* Channels 0 to 7 */
+	u32_t	req_chan_map_hi;	/* Channels 8 to 15 */
+	/* Actual channel map that the driver could configure */
+	u32_t	act_chan_map_lo;	/* Channels 0 to 7 */
+	u32_t	act_chan_map_hi;	/* Channels 8 to 15 */
+	/* requested number of channels */
+	u8_t	req_num_chan;
+	/* Actual number of channels that the driver could configure */
+	u8_t	act_num_chan;
+	/* requested number of streams for each channel */
+	u8_t	req_num_streams;
+	/* Actual number of streams that the driver could configure */
+	u8_t	act_num_streams;
+};
+
+/**
+ * Input configuration structure for the DMIC configuration API
+ */
+struct dmic_cfg {
+	struct pdm_io_cfg io;
+	/*
+	 * Array of pcm_stream_cfg for application to provide
+	 * configuration for each stream
+	 */
+	struct pcm_stream_cfg *streams;
+	struct pdm_chan_cfg channel;
+};
+
+/**
+ * Function pointers for the DMIC driver operations
+ */
+struct _dmic_ops {
+	int (*configure)(struct device *dev, struct dmic_cfg *config);
+	int (*trigger)(struct device *dev, enum dmic_trigger cmd);
+	int (*read)(struct device *dev, u8_t stream, void **buffer,
```

Comment:
Read is fine but I have one comment.
In typical case application would have to create a thread to wait for data. If there is more then one stream, there will have to be one dedicated thread for one stream. This is problematic as each thread has to have its own stack causing memory waste.

I don't think there is a generic API for poll in Zephyr but maybe we can start here. Can we add a function to wait for any stream to be ready (*wait_for_streams(..*dev, u8_t *streams, size_t *count)), that would return when data is available in any stream. Arguments: stream - array to hold ids of streams ready, count - number entries in array when called, number of stream ready when returned back. This way thread would call it, and only upon its return would call read on the streams that are ready.

@carlescufi , @pizi-nordic This look like another example when poll similar to linux would be useful.

Reply:
> If there is more then one stream, there will have to be one dedicated thread for one stream. This is problematic as each thread has to have its own stack causing memory waste.

It's not problematic. The streams are application specific (voice calling application will require lower sample rate compared to a high quality audio recording for reproduction purposes). So having a read with stream id that the respective application can read using it's own thread is going to be use case.

Reply:
@sathishkuttan I don't agree. Everything depends on the application. It may be that one would like to read two streams on SoC with low RAM and (with this API) will be forced to waste memory for additional thread's stack.
If polling (in whatever form) is present, application can use it (to read many streams from single thread) or don't (create a thread or even an application for one stream). But it is an application choice.
Anyway this is a general problem in Zephyr.

Reply:
@pdunaj You don't have to agree. You are entitled to your opinion.

Reply:
@sathishkuttan Thanks ;)

Reply:
@pdunaj You are welcome :)
By the way, this API does not force to waste memory. In my actual implementation, I am reading 2 streams from one thread -- one from this DMIC and another from I2S. The thread blocks on DMIC read and once the data is read from DMIC, the I2S data is read without a wait. Through a proper hardware scheduling, I ensure that the I2S data is ready when the DMIC data is ready.
While in this case, it is between DMIC and I2S, the same concept can be used for 2 DMIC streams.